### ISSUE TITLE
Why doesn't the second constructor for a parent class violate the principle of Tell Don't Ask? 

### ISSUE THREAD

**Issue by Jeremythong2002** [nus-cs2030/2324-s2#52]
# Description

In lecture 1, it was mentioned that we should not make use of getter methods (Tell Don’t Ask) as it would defeat the purpose of making this attributes private. However, lecture 3, Prof Henry used a second constructor which essentially acts like a getter method to access the radius of FilledCircle from the parent class. 

Could I check what is the guideline for using a getter and why is it a better practice to create a second constructor rather than simply using protected attribute? 

## Topic:

OOP

**Comment by hwkchia** [comment]
"Tell Don't Ask" literally means don't get the radius value from the parent class and make use of it.  In using the second constructor, no radius is passed.  Is the circle (or rather `this`) that is passed.  Specifically,

`FillCircle(color)` makes use of `FilledCircle(this, color)` which makes use of `Circle(this)`.  Notice there is no indication of `radius`?  `FilledCircle` is telling `Circle` to construct another circle and `Circle` deals with the radius assignment.

**Comment by chinwn** [comment]
One of the reasons for avoiding getter methods is that they reveal implementation details, e.g. how Circle is being implemented via radius. The implementor of Circle may wish to change from _radius_ to _diameter_ as its new implementation of Circle. With abstraction in place via private modifier, such a change can be conducted entirely within the Circle class since the 2nd constructor method is still within this Class. Hence, FilledCircle as client of Circle is completely unaffected since it is only making use of the service from the 2nd constructor method, and did not have direct access to radius as its implementation detail.

In summary, implementation details of classes are often changed by developer during maintenance and improvement phases. However, functionality of classes are quite stable, though they may still change from time-to-time depending also on the requirements from clients. 

Tell-don't-Ask principle is thus to help enforce abstraction wall to isolate client from unnecessary implementation details that may be _freely_ changed by developers from time-to-time without having to inform clients.

**Comment by hwkchia** [comment]
Contributions included in Lecture 03.
